dap-docker.el

```diff
@@ -0,0 +1,253 @@
+             (server-subtype (gethash 'subtype server-info)))
+      (if (equal server-subtype "container")
+          (gethash 'name server-info))
+    (user-error "No debug container names are specified, you cannot use single container both for a language server and a debugging server")))
+
+(defun dap-docker--get-server-image-name (config)
+  "Get the server image name"
+  (-if-let* ((server-info (gethash 'debug config))
+             (server-subtype (gethash 'subtype server-info)))
+      (if (equal server-subtype "image")
+          (gethash 'name server-info))
+    (lsp-docker-get-server-image-name config)))
+
+(defun dap-docker--get-server-id (config)
+  "Get the server id"
+  (let ((server-info (gethash 'debug config)))
+    (if (stringp (gethash 'server server-info))
+        (intern (gethash 'server server-info))
+      (gethash 'server server-info))))
+
+(defun dap-docker--get-path-mappings (config)
+  "Get the server path mappings"
+  (lsp-docker-get-path-mappings config))
+
+(defun dap-docker--get-launch-command (config)
+  "Get the server launch command"
+  (let ((server-info (gethash 'debug config)))
+    (gethash 'launch_command server-info)))
+
+(defun dap-docker--check-server-type-subtype (supported-server-types-subtypes server-type-subtype)
+  "Verify that the combination of server (type . subtype) is supported by the current implementation"
+  (if (not server-type-subtype)
+      (user-error "No server type and subtype specified!"))
+  (if (ht-find (lambda (type subtypes)
+                 (let ((server-type (car server-type-subtype))
+                       (server-subtype (cdr server-type-subtype)))
+                   (if (and (equal server-type type) (-contains? subtypes server-subtype))
+                       t)))
+               supported-server-types-subtypes)
+      server-type-subtype
+    (user-error "No compatible server type and subtype found!")))
+
+(defun dap-docker--check-path-mappings (path-mappings)
+  "Verify that specified path mappings are all inside the project directory"
+  (--all? (or (f-descendant-of? (f-canonical (car it)) (f-canonical (lsp-workspace-root)))
+              (f-same? (f-canonical (car it)) (f-canonical (lsp-workspace-root))))
+          path-mappings))
+
+(defun dap-docker--verify-path-mappings-against-container (path-mappings container-name)
+  "Verify that specified path mappings are all included in container's path mappings"
+  (--all? (let ((source (car it))
+                (destination (cdr it)))
+            (-any? (lambda (mapping)
+                     (and (f-same? source (car mapping))
+                          (equal destination (cdr mapping))))
+                   (dap-docker--get-path-mappings-from-container container-name)))
+          path-mappings))
+
+(defun dap-docker--get-path-mappings-from-container (container-name)
+  "Get path mappings from a container"
+  (-let ((
+          (inspection-command-program . inspection-command-arguments)
```

Comment:
nit: on the same line(here and through the review)

Reply:
Fixed in f69dbb1